flocker/common/test/test_thread.py

```diff
@@ -0,0 +1,179 @@
+        arguments.
+        """
+        self.calls.append((a, b, c))
+        return a + b + c
+
+
+@auto_threaded(IStub, "reactor", "provider", "threadpool")
+class AsyncSpy(PRecord):
+    """
+    An automatically asynchronous version of ``Spy``.
+    """
+    reactor = field()
+    provider = field()
+    threadpool = field()
+
+
+class NonThreadPool(object):
+    """
+    A stand-in for ``twisted.python.threadpool.ThreadPool`` so that the
+    majority of the test suite does not need to use multithreading.
+
+    This implementation takes the function call which is meant to run in a
+    thread pool and runs it synchronously in the calling thread.
+    """
+    def callInThreadWithCallback(self, onResult, func, *args, **kw):
+        try:
+            result = func(*args, **kw)
+        except:
+            onResult(False, Failure())
+        else:
+            onResult(True, result)
+
+
+class NonReactor(object):
+    """
+    A stand-in for ``twisted.internet.reactor`` which fits into the execution
+    model defined by ``NonThreadPool``.
+    """
+    def callFromThread(self, f, *args, **kwargs):
+        f(*args, **kwargs)
+
+
+class AutoThreadedTests(SynchronousTestCase):
+    """
+    Tests for ``flocker.common.auto_threaded``.
+    """
+    def setUp(self):
+        self.reactor = NonReactor()
+        self.threadpool = NonThreadPool()
+        # Some unique objects that support ``+``
+        self.a = [object()]
+        self.b = [object()]
+        self.c = [object()]
+        self.spy = Spy()
+        self.async_spy = AsyncSpy(
+            reactor=self.reactor, threadpool=self.threadpool, provider=self.spy
+        )
+
+    def test_called_with_same_arguments(self):
+        """
+        A class can be decorated with ``auto_threaded`` with an interface and a
+        provider of that interface.  When methods from the interface are called
+        on instances of the class, the corresponding method is called with the
+        same arguments on the provider of the interface.
+        """
+        self.async_spy.method(self.a, c=self.c, b=self.b)
+        self.assertEqual([(self.a, self.b, self.c)], self.spy.calls)
+
+    def test_success_result(self):
+        """
+        When methods from the interface are called on instances of the class, a
+        ``Deferred`` is returned which fires with the result of the call to the
+        corresponding method on the provider of the interface.
+        """
+        args = (self.a, self.b, self.c)
+        result = self.successResultOf(self.async_spy.method(*args))
+        self.assertEqual(self.spy.method(*args), result)
+
+    def test_failure_result(self):
+        """
+        When the corresponding method on the provider of the interface raises
+        an exception, the ``Deferred`` fires with a ``Failure`` representing
+        that exception.
+        """
+        self.failureResultOf(
+            self.async_spy.method(None, None, None),
+            TypeError
```

Comment:
`TypeError` seems a bit generic. Could be caused by a mistake in the implementation, right? 
Maybe raise and check for a `Spy` specific exception type?